docs/proposals-accepted/202304-query-path-tenancy.md

```diff
@@ -0,0 +1,80 @@
+
+The current lack of tenancy awareness in Thanos' query path makes it impossible to investigate issues related to multi-tenancy without the use of external tools or proxies. For example, it's impossible to determine which tenants are experiencing high latency or high error rates.
+
+## Goals
+
+* Allow the query path components to be configurable to identify tenants, opening the way to the implementation per-tenant features on the query path. These features include, but aren't limited to, the following:
+  * Per-tenant observability.
+  * Per-tenant settings. For example, having different limits per tenant, which is a common request.
+  * Enforce presence of one or more tenant labels in queries.
+
+### Audience
+
+Any team running Thanos in a multi-tenant environment. For example, a team running a monitoring-as-a-service platfor. For example, a team running a monitoring-as-a-service platform.
+
+## Non-Goals
+
+* Add multi-tenancy to Thanos Ruler. 
+* Implement cross-tenant querying. It poses the question about how to track metrics of a multi-tenant query, probably requiring new logic for query splitting and/or separation of the concepts of query initiator tenant and query target tenants. To exemplify: if tenant A executes a query touching their own data and tenant B data (imagine a label matcher `tenant_id=~"A|B"`), how does Thanos export a request duration metric about this query? This is out of scope and an specific proposal can be created for it.
+
+## How
+
+* Implement a mechanism to allow incoming requests to specify the tenant being queried using an HTTP header. This applies to all components in the query path. We follow the example of Thanos Receive, which uses the `--receive.tenant-header="THANOS-TENANT"` flag to configure the tenant header, adapting its name to each component. So in the Querier, the flag name will be `--querier.tenant-header`, in the Store Gateway it'll be `--store.tenant-header`, and so on.
+* The tenant header value from a given request should travel downstream to all the components being called, so that it can be added to their metrics, traces, and logs without requiring duplicated/extra work to re-parse the query. This also applies to gRPC calls.
+* Implement a command line argument in the Querier component that will indicate with label name that should be used to enforce tenancy. Again, we follow the example of Thanos Receive, which uses the `--receive.tenant-label-name="tenant_id"` flag to achieve this, and adapt its name just as described on the first point (changing the `receive` part to `querier`, `store`, etc). The default behavior, when the flag as an empty string as value, is to not identify tenants.
+  * The label verification and enforcement should be done by reusing prom-label-proxy's [Enforce.EnforceMatchers](https://github.com/prometheus-community/prom-label-proxy/blob/main/injectproxy/enforce.go#L141). There's no reason to (re)implement something specific and special for Thanos.
```

Comment:
I don't think we should be modifying incoming promql expressions since that can have unexpected side effects, especially with functions like `absent`. 

I would suggest enforcing the tenant in the storage querier by directly adding a separate label matcher: https://github.com/thanos-io/thanos/blob/main/pkg/query/querier.go.

Reply:
Ah just realized that enforce matchers does exactly that. Using #4141 as a starting point makes sense 👍